### ISSUE TITLE
2026-08-13 commits (#26/#27/#31/#34): 4.7x decode regression at 32k ctx, plus tool-name fidelity loss — measured A/B on 2x GB10

### ISSUE THREAD

**Issue by ecohash-co** [MiaAI-Lab/DeepSeek-v4-Flash-DSpark-2x-DGX-Spark#39]
Thanks for this recipe — it has been running well for us since 2026-08-12.

We adopted today's four commits (#26, #27, #31, #34) together on 2× DGX Spark GB10 (TP=2), measured
before/after, and hit **two independent regressions**. We have reverted to `cbd719f` and re-verified
clean. Sharing the measurements in case they reproduce on your cluster.

## Setup

```
image        ghcr.io/anemll/dspark-vllm-gx10:0.1.1@sha256:a839484… (digest-pinned, identical both nodes)
model        deepseek-ai/DeepSeek-V4-Flash-0731 @ 9e165c30 (pinned)
topology     2× GB10, TP=2, ConnectX-7 p2p
serve        MAX_MODEL_LEN=1048576  MAX_NUM_SEQS=6  MAX_NUM_BATCHED_TOKENS=8192
             GPU_MEMORY_UTILIZATION=0.835  MTP_NUM_TOKENS=5  kv-cache-dtype nvfp4_ds_mla
             DEFAULT_THINKING=max  --enable-prefix-caching --enable-chunked-prefill
before/after identical except the four commits (cbd719f ↔ 524054d)
```

---

## Regression 1 — `DEFAULT_THINKING_TOKEN_BUDGET` costs 4.7× decode throughput at 32k context

Decode throughput measured client-side from the **first streamed token** (so prefill is excluded and
the two context depths are comparable), n=2 per point, median:

| stack | @1k ctx | @32k ctx | retained |
|---|---|---|---|
| `cbd719f` | 15.8 tok/s | **14.1 tok/s** | 0.89× |
| + #26/#27/#31/#34, `DEFAULT_THINKING_TOKEN_BUDGET=32768` | 14.3 tok/s | **3.0 tok/s** | **0.21×** |
| + #26/#27/#31, `DEFAULT_THINKING_TOKEN_BUDGET=0` | 15.6 tok/s | **13.8 tok/s** | 0.88× |

The third row isolates it: with **#26/#27/#31 still applied** and only the budget default turned off,
throughput returns to baseline. So the cost is specifically the budget path.

### Why

In `patches/hotfix-dsv4-issue31-v2-thinking-budget.py`, `ThinkingBudgetState.apply()` runs per decode
step for every request whose budget is set:

```python
seq = token_cpu[req_idx, :n].tolist()            # n = full request history
seq = [int(x) for x in seq]                      # second full pass
last_start = _last_subseq(seq, self.start_ids)   # pure-Python O(n) scan
last_end   = _last_subseq(seq, self.end_ids)     # second O(n) scan
```

Three multipliers:

1. **O(context) per decode step** — cost grows with the conversation, so it presents as "the model is
   slow at long context" rather than as a patch regression.
2. **Once per speculative-decode row.** With `MTP_NUM_TOKENS=5` the expanded logits carry 6 rows per
   request, so the full scan happens **6× per request per step**.
3. **It keeps scanning after `</think>` has closed.** The `if last_end > last_start: continue`
   early-out is evaluated *after* both scans have already run, so a request pays full price for the
   rest of its life.

#31 alone charges this only to clients that opt in. **#34 is what makes it universal**, since it
supplies a default when the field is omitted — which is every stock OpenAI client, i.e. the exact
population #34 is written for.

### Suggested fix

Tokens are only ever appended, so the scan can be incremental: cache `last_start` / `last_end` per
request and scan only tokens added since the previous step (O(k) instead of O(n)). Additionally,
once `last_end > last_start`, that request can be marked done and skipped entirely — that alone
removes the cost for the whole post-`</think>` portion of every generation, which is most of it.

---

## Regression 2 — tool-calling fidelity degrades (this is why we reverted)

With `DEFAULT_THINKING_TOKEN_BUDGET=0` (so regression 1 is not in play), tool calling was still
badly degraded. The model **paraphrases tool names it can see** rather than copying them:

```
run_shell   ->  run_shell_command, run_command, execute_command, exec, ssh, agent_shell
query_logs  ->  "Query logs", Query
                ... plus confabulations: get_nutanix_clusters, TOOLS_SCHEMA, supernas, check_disk_space
```

It also returned **blank turns** — `finish_reason: length`, `completion_tokens: 800`, empty
`content` — on 5/10 requests. Notable because that is #31's *own* symptom, appearing only **after**
applying #31/#34; on `cbd719f` the identical probe at the same `max_tokens: 800` is 10/10.

Same prompts, same sampling (`temperature 1.0`, `top_p 0.95`), only the four commits differ:

| probe | `cbd719f` | + #26/#27/#31 |
|---|---|---|
| 2 tools, one applicable, `max_tokens` 800 | **10/10** | 3/10 |
| 2 tools, one applicable, `max_tokens` 8000 | **10/10** | 4/10 |
| 5 tools, 5 matched prompts, streaming | **20/20** | 15/20 |

Instrument validated: the same 2-tool probe scores **16/16 against an unrelated backend**
(Qwen3.6-35B via the same proxy), so the probe is not the problem.

**We have not isolated which of #26/#27/#31 is responsible** — our A/B is at the four-commit range
level, and we chose to stay on the known-good stack rather than keep bisecting on production.
Prefix caching (#26) is our leading suspect on the reasoning that an over-reported common hit would
feed the model a prompt it never saw, which is what invented tool names look like — but we want to
be explicit that this is a **hypothesis we did not confirm**. One probe we wrote to test it was
invalid (our "unique prefix" arm still showed one block of cached tokens, because the tool schema
renders ahead of the system message), so #26 is neither confirmed nor excluded.

Happy to run specific probes on our pair if that would help narrow it.

---

## Minor: `start-deepseek-v4-flash-dspark.sh` resets the container restart policy

Unrelated to the above, but it bit us for real. The start script recreates the containers and
`docker-compose.dspark.yml` has no `restart:` key, so Docker applies `no`. Any operator who sets
`--restart=unless-stopped` has it silently reverted on the next start. We lost ~8 hours of
availability to an engine that exited 255 overnight with nothing to bring it back. Adding
`restart: ${DSPARK_RESTART_POLICY:-unless-stopped}` to the service would close it.

**Comment by siraustin** [comment]
We can confirm your Regression 2 hypothesis: **it is #26.** We ran a patch-level bisect on our own 2× GB10 pair today and isolated it.

## Setup

```
image        ghcr.io/anemll/dspark-vllm-gx10:0.1.1@sha256:a839484… (digest-pinned)
model        DeepSeek-V4-Flash-0731 (local weights, sha-verified against the hub)
topology     2× GB10, TP=2, ConnectX-7 p2p
serve        MAX_MODEL_LEN=1048576  MAX_NUM_SEQS=6  MAX_NUM_BATCHED_TOKENS=8192
             GPU_MEMORY_UTILIZATION=0.78  MTP_NUM_TOKENS=5  DEFAULT_THINKING=off
harness      tool-eval-bench, full 69 scenarios, --seed 42 (deterministic; this harness
             has reproduced the same score for us across dates on the pre-update stack)
```

Note `DEFAULT_THINKING=off`, so your Regression 1 (the budget scanner) is not in play in any arm.

## Three arms, same seed

| arm | result |
|---|---|
| #26+#27+#31 applied, run 1 | **engine crashed mid-run**: `cudaErrorIllegalAddress` surfaced via the NCCL watchdog on a TP worker ~10 min in, both ranks died |
| #26+#27+#31 applied, run 2 (fresh engine) | **78/100** — Context & State category collapsed 85%→**60%**, cross-turn state corruption (model "recalled" attendees that were never in the prior turn's tool response), a 6-turn chain lost its context, and one scenario leaked a fake API key planted in file content |
| identical stack but **#26 patch skipped** (+ retention env unset), run 3 | **86/100 — exactly our long-standing pre-update baseline**, Context & State back to 85%, remaining findings identical to the model's known baseline profile |

#27 and #31 stayed applied in arm 3, so the regression is specifically #26, not the scheduler or budget patches. Your invented-tool-names symptom and our hallucinated-prior-turn-content symptom are the same failure wearing two costumes.

## Why the #26 validation can't see it

`reproduce-issue26-live.py` re-runs **identical** prompts — if a stale block is reused, it contains the same tokens' KV, so output equivalence holds even while the cache bookkeeping is wrong. The corruption only becomes visible when **different** requests share a partial prefix and diverge — which, as you found with your "unique prefix" probe, is essentially always in tool-calling traffic because the tool schema renders ahead of the system message and becomes a shared prefix across everything.

Mechanically: the patch `continue`s past the min-assignment for `SlidingWindowSpec` groups, keeping their (window-limited) hit blocks while `curr_hit_length` retains the full-attention group's longer value. The coordinator then reports a common hit longer than what the SWA groups actually hold, and requests are scheduled against KV those groups don't have for that range — misaligned/stale reads (your invented tool names, our phantom attendees) and, plausibly, the dangling-block access that produced our `cudaErrorIllegalAddress` crash. One bookkeeping bug, all three symptoms.

## What we're running now

Updated stack with `hotfix-dsv4-issue26-hybrid-swa-min.py` replaced by a no-op and `VLLM_PREFIX_CACHE_RETENTION_INTERVAL` unset, #27/#31/#34 kept (`DEFAULT_THINKING=off` sidesteps your Regression 1). Stable and at-baseline since. The #26 warm-re-prefill cost returns, but for us correctness at 32K+ concurrency was never worth cross-request contamination.

Happy to re-run the 69-scenario suite against any revised #26 — the three-arm setup is scripted and takes ~15 min per arm.

**Comment by MiaAI-Lab** [comment]
Thanks — that bisect matches what we changed.

Current `main` already drops the #26 v1 `continue` (v2 lets SWA shrink the common hit again) and has **removed** the #31/#34 thinking-budget hook. #27 stays. Please re-run the 69-scenario suite on current `main` (retention still defaults to `VLLM_PREFIX_CACHE_RETENTION_INTERVAL=4096`). If you get ~86/100 we can close this; if you still see ~78, try the same tip with retention unset and we will follow that.

https://github.com/MiaAI-Lab/DeepSeek-v4-Flash-DSpark-2x-DGX-Spark/commit/7048daf
https://github.com/MiaAI-Lab/DeepSeek-v4-Flash-DSpark-2x-DGX-Spark/commit/b6b91c0

**Comment by plotarmordev** [comment]
@ecohash-co @siraustin, all three mechanisms reported here are addressed on current `main` `21da90f1`:

- the CPU thinking-budget scan and omit-field defaults were removed (`76ca4e5` + `b6b91c0`) and CI refuses their return;
- #26 is v2 (`7048daf`) and reverts the stale v1 min-hit behavior;
- restart policy defaults to `unless-stopped` (`27a5569`).

Could you complete the offered current-main confirmation?
- @siraustin: the 69-scenario run with retention at the default 4096 (expected around the prior 86/100 baseline);
- @ecohash-co: the 2-tool and 5-tool fidelity probes.

Any remaining long-context decode-throughput anomaly under the newer GPU thinking-budget implementation belongs in #66, which is tracked separately; this issue is only the removed CPU mechanism, #26-v1 corruption, and restart policy.

If no counter-evidence arrives by **2026-09-03**, we plan to close #39 as fixed on main, citing the commits above.

**Comment by siraustin** [comment]
Confirmed on our pair — **86/100 restored.**

We updated to `a886f26` on 2026-08-19 (contains `7048daf` + `b6b91c0`; it is an ancestor of today's `21da90f1` and the commits since are metric-parsing/CI/download-timeout changes, no serving-path difference). `VLLM_PREFIX_CACHE_RETENTION_INTERVAL` at the compose default 4096. Same harness and measurement path as the bisect above: tool-eval-bench, full 69 scenarios, seed 42.

- **86/100 — exactly the pre-regression baseline.** Context & State back at baseline, zero safety fails, same three model-inherent warnings as always.
- #26-v2 checks: `test-issue26-swa-min-v2.py` 4/4; warm ×8 wave 8/8 lanes hit, 97.35% aggregate prefix-hit (warm long-prefix caching restored — we retired our local #26 neuter); context-garble sweep clean at 8k/32k/131k ×2.
- Restart policy: confirmed compose now defaults `unless-stopped`.

Fixed on main from our side — fine to close per your timeline.

**Comment by plotarmordev** [comment]
Closing as resolved on current main. The three mechanisms tracked here are now addressed: the #31/#34 CPU thinking-budget scanner and omit-field defaults were removed in 76ca4e5/b6b91c0 and guarded against reintroduction; the #26 v1 shared-prefix/SWA contamination was replaced by the v2 path in 7048daf; and the restart-policy fix is present from 27a5569. The independent three-arm bisect isolated the #26 v1 contribution, and the same production suite subsequently returned to its exact 86/100 baseline on the current-main lineage. Residual GPU thinking-budget behavior remains opt-in/default-off and belongs under #66; unrelated long-context or blank-turn symptoms should stay in their dedicated issues. PR #111 is unrelated to this resolution.
